File: app/src/main/java/com/stedi/randomimagegenerator/app/presenter/impl/ChooseEffectPresenterImpl.java
package com.stedi.randomimagegenerator.app.presenter.impl;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.stedi.randomimagegenerator.app.model.data.GeneratorType;
import com.stedi.randomimagegenerator.app.model.data.PendingPreset;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.base.EffectGeneratorParams;
import com.stedi.randomimagegenerator.app.model.data.generatorparams.base.GeneratorParams;
import com.stedi.randomimagegenerator.app.other.logger.Logger;
import com.stedi.randomimagegenerator.app.presenter.interfaces.ChooseEffectPresenter;

public class ChooseEffectPresenterImpl implements ChooseEffectPresenter {
    private final PendingPreset pendingPreset;
    private final Logger logger;

    private UIImpl ui;

    public ChooseEffectPresenterImpl(@NonNull PendingPreset pendingPreset, @NonNull Logger logger) {
        if (pendingPreset.getCandidate() == null)
            throw new IllegalStateException("pending preset candidate must not be null");
        this.pendingPreset = pendingPreset;
        this.logger = logger;
    }

    @Override
    public void getEffectTypes() {
        GeneratorParams currentParams = pendingPreset.getCandidate().getGeneratorParams();
        GeneratorType effectType;
        GeneratorType targetType;
        if (currentParams instanceof EffectGeneratorParams) {
            effectType = currentParams.getType();
            targetType = ((EffectGeneratorParams) currentParams).getTarget().getType();
        } else {
            effectType = null;
            targetType = currentParams.getType();
        }
        ui.showTypes(GeneratorType.effectTypes(), effectType, targetType);
    }

    @Override
    public void chooseEffectType(@Nullable GeneratorType effectType) {
        GeneratorParams newParams;
        GeneratorParams prevParams = pendingPreset.getCandidate().getGeneratorParams();
        if (prevParams instanceof EffectGeneratorParams) {
            prevParams = ((EffectGeneratorParams) prevParams).getTarget();
        }
        if (effectType != null) {
            newParams = GeneratorParams.Companion.createDefaultEffectParams(effectType, prevParams);
        } else {
            newParams = prevParams;
        }
        pendingPreset.getCandidate().setGeneratorParams(newParams);
        logger.log(this, "chooseGeneratorType result = " + pendingPreset.getCandidate().getGeneratorParams());
    }

    @Override
    public void onAttach(@NonNull UIImpl ui) {
        this.ui = ui;
    }

    @Override
    public void onDetach() {
        this.ui = null;
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/presenter/impl/ChooseEffectPresenterImpl.kt
package com.stedi.randomimagegenerator.app.presenter.impl

import com.stedi.randomimagegenerator.app.model.data.GeneratorType
import com.stedi.randomimagegenerator.app.model.data.PendingPreset
import com.stedi.randomimagegenerator.app.model.data.Preset
import com.stedi.randomimagegenerator.app.model.data.generatorparams.base.EffectGeneratorParams
import com.stedi.randomimagegenerator.app.model.data.generatorparams.base.GeneratorParams
import com.stedi.randomimagegenerator.app.presenter.interfaces.ChooseEffectPresenter
import timber.log.Timber
import javax.inject.Inject

class ChooseEffectPresenterImpl @Inject constructor(
        private val pendingPreset: PendingPreset) : ChooseEffectPresenter {

    private val candidate: Preset
        get() = pendingPreset.getCandidate()
                ?: throw IllegalStateException("pending preset candidate must not be null")

    private var ui: ChooseEffectPresenter.UIImpl? = null

    override fun onAttach(ui: ChooseEffectPresenter.UIImpl) {
        this.ui = ui
    }

    override fun onDetach() {
        this.ui = null
    }

    override fun getEffectTypes() {
        val currentParams = candidate.getGeneratorParams()
        val effectType: GeneratorType?
        val targetType: GeneratorType
        if (currentParams is EffectGeneratorParams) {
            effectType = currentParams.getType()
            targetType = currentParams.target.getType()
        } else {
            effectType = null
            targetType = currentParams.getType()
        }
        ui?.showTypes(GeneratorType.EFFECT_TYPES, effectType, targetType)
    }

    override fun chooseEffectType(effectType: GeneratorType?) {
        val newParams: GeneratorParams
        var prevParams = candidate.getGeneratorParams()
        if (prevParams is EffectGeneratorParams) {
            prevParams = prevParams.target
        }
        if (effectType != null) {
            newParams = GeneratorParams.createDefaultEffectParams(effectType, prevParams)
        } else {
            newParams = prevParams
        }
        candidate.setGeneratorParams(newParams)
        Timber.d("chooseGeneratorType result = ${candidate.getGeneratorParams()}")
    }
}